[AIRFLOW-1002] Add ability to clean all dependencies of removed DAG

sekikn opened the pull request:
Dear Airflow maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!


### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title.
    - https://issues.apache.org/jira/browse/AIRFLOW-1002

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:

After removing a dag file, there is no way to clean database
except for removing corresponding records directly for now.
This PR enables user to do this via command line.


### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

CliTests.test_cli_remove_dag


### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

bolkedebruin commented on airflow/bin/cli.py:
```diff
@@ -208,6 +208,28 @@ def _tabulate(pools):
+    :param args:
+    :return:
+    """
+    print("DB: " + repr(settings.engine.url))
```
Why are you printing this? In case of the JSONClient it is not relevant.

bolkedebruin commented on airflow/www/api/experimental/endpoints.py:
```diff
@@ -234,3 +235,20 @@ def delete_pool(name):
+    except Exception as err:
+        _log.error(err)
+        response = jsonify(error="{}".format(err))
+        response.status_code = 404
```
How do you know it is a 404? I could be a 500 error. Please make your exception handling less broad

bolkedebruin commented on tests/core.py:
```diff
@@ -1409,6 +1409,23 @@ def test_pool_no_args(self):
+        session = settings.Session()
+        session.add(models.DagModel(dag_id=key))
+        session.commit()
+        cli.remove_dag(self.parser.parse_args([
```
You are not checking if anything was removed

bolkedebruin commented on airflow/api/common/experimental/remove_dag.py:
```diff
@@ -0,0 +1,42 @@
+
+    session = settings.Session()
+    DM = models.DagModel
+    dag = session.query(DM).filter(DM.dag_id==dag_id).first()
```
You probably want to check if the dag exists in the DagBag as well. If it does, it will get re-added which could lead to some race conditions.

bolkedebruin commented on airflow/api/common/experimental/remove_dag.py:
```diff
@@ -0,0 +1,42 @@
+            cond = or_(m.dag_id==dag_id, m.dag_id.like(dag_id + ".%"))
+            count += session.query(m).filter(cond).delete(synchronize_session='fetch')
+
+    if dag.is_subdag:
```
Why only remove task_instances for SubDags? Why not remove dag_runs too?

duynguyenhoang commented on airflow/www/api/experimental/endpoints.py:
```diff
@@ -234,3 +235,20 @@ def delete_pool(name):
+
+
+@csrf.exempt
+@api_experimental.route('/dags/<string:dag_id>/remove_dag', methods=['POST'])
```
For this endpoint, I suggest to use 'DELETE' method and also remove `/remove_dag`. The route will become
`@api_experimental.route('/dags/<string:dag_id>', methods=['DELETE'])`

duynguyenhoang commented on airflow/www/api/experimental/endpoints.py:
```diff
@@ -234,3 +235,20 @@ def delete_pool(name):
+        response = jsonify(error="{}".format(e))
+        response.status_code = 500
+        return response
+    return jsonify(message="Removed {} record(s)".format(count))
```
It will be better if we can have `count` inside json. Like below example

```
{
    "message": "Removed 10 record(s)",
    "count": 10,
}
```

sekikn committed:
[AIRFLOW-1002] Add ability to clean all dependencies of removed DAG

After removing a dag file, there is no way to clean database
except for removing corresponding records directly for now.
This PR enables user to do this via command line.
